client/file-manager/src/rocksdb.rs

```diff
@@ -0,0 +1,1023 @@
+
+        // We Read all the chunks to prove from the trie.
+        // This is step is required to actually record the proof.
+        let mut chunks = Vec::new();
+        for chunk_id in chunk_ids {
+            let chunk: Option<Vec<u8>> = trie
+                .get(&chunk_id.as_trie_key())
+                .map_err(|_| FileStorageError::FailedToGetFileChunk)?;
+
+            let chunk = chunk.ok_or(FileStorageError::FileChunkDoesNotExist)?;
+            chunks.push((*chunk_id, chunk));
+        }
+        // Drop the `trie_recorder` to release the `recorder`
+        drop(trie_recorder);
+
+        // Generate proof
+        let proof = recorder
+            .drain_storage_proof()
+            .to_compact_proof::<T::Hash>(self.root)
+            .map_err(|_| FileStorageError::FailedToGenerateCompactProof)?;
+
+        Ok(FileProof {
+            proof: proof.into(),
+            fingerprint: self.get_root().as_ref().into(),
+        })
+    }
+
+    // TODO: make it accept a list of chunks to be retrieved
+    fn get_chunk(&self, chunk_id: &ChunkId) -> Result<Chunk, FileStorageError> {
+        let db = self.as_hash_db();
+        let trie = TrieDBBuilder::<T>::new(&db, &self.root).build();
+
+        trie.get(&chunk_id.as_trie_key())
+            .map_err(|_| FileStorageError::FailedToGetFileChunk)?
+            .ok_or(FileStorageError::FileChunkDoesNotExist)
+    }
+
+    // TODO: make it accept a list of chunks to be written
+    fn write_chunk(
+        &mut self,
+        chunk_id: &ChunkId,
+        data: &Chunk,
+    ) -> Result<(), FileStorageWriteError> {
+        let mut current_root = self.root;
+        let db = self.as_hash_db_mut();
+        let mut trie = TrieDBMutBuilder::<T>::from_existing(db, &mut current_root).build();
+
+        // Check that we don't have a chunk already stored.
+        if trie.contains(&chunk_id.as_trie_key()).map_err(|e| {
+            error!(target: LOG_TARGET, "{}", e);
+            FileStorageWriteError::FailedToGetFileChunk
+        })? {
+            return Err(FileStorageWriteError::FileChunkAlreadyExists);
+        }
+
+        // Insert the chunk into the file trie.
+        trie.insert(&chunk_id.as_trie_key(), &data).map_err(|e| {
+            error!(target: LOG_TARGET, "{}", e);
+            FileStorageWriteError::FailedToInsertFileChunk
+        })?;
+
+        // get new root after trie modifications
+        let new_root = *trie.root();
+        // drop trie to commit to underlying db and release `self`
+        drop(trie);
+
+        // TODO: improve error handling
+        // Commit the changes to disk.
+        self.commit(new_root).map_err(|e| {
+            error!(target: LOG_TARGET, "Failed to commit changes to persistent storage: {}", e);
+            FileStorageWriteError::FailedToPersistChanges
+        })?;
+
+        Ok(())
+    }
+
+    // Deletes itself from the underlying db.
+    fn delete(&mut self, chunk_count: u64) -> Result<(), FileStorageWriteError> {
```

Comment:
Why do you need to pass `chunk_count` here? It would be more ergonomic to not force callers have to get the number of chunks beforehand.

Reply:
The caller here would be `FileStorage`, which gets the chunk count directly from the `FileMetadata` that it stores. It seemed better than iterating the Trie inside `FileDataTrie` and then removing each chunk.